File: bot/src/command/PlanCommand.ts
import Command from "./Command";
import {VkMessage} from "../service/VkMessagesService";
import ServiceError from "../ServiceError";

export default class PlanCommand extends Command {
    canYouHandleThisCommand(command: string, message: VkMessage): boolean {
        return command == "plan";
    }

    getCommandShortUsage(): string {
        return "/sweet plan (...)";
    }

    chatAdminOnly(peerId: number): boolean {
        return false;
    }

    async handle(command: string, rawArguments: string, message: VkMessage): Promise<void> {
        const { vkMessagesService } = this.context;

        const subCommand = rawArguments.split(" ")[0];
        if (!subCommand) {
            await this.handleDefault(message);
        } else if (subCommand == 'list') {
            await this.handleList(message);
        } else if (subCommand == 'help') {
            await vkMessagesService.send(message.peerId, this.getUsage());
        } else {
            await vkMessagesService.send(message.peerId, this.getUsage());
        }
    }

    private async handleDefault(message: VkMessage): Promise<void> {
        const { vkMessagesService } = this.context;
        const user = await this.context.vkUsersService.getUser(message.fromId);
        if (!user) {
            throw new ServiceError("user not found");
        }
        const plan = await this.context.usagePlanService.getUsagePlanOrDefault(user.usagePlanId);
        const planTitle = plan ? plan.title : "(no plan)";
        let response = ``;
        response += `⚡ Sweetie AI ${planTitle}\n`;
        response += `🪙 Кредиты: ${user.credits}`;
        if (plan) {
            response += `/${plan.maxCredits}`;
        }
        response += `\n`;
        if (plan) {
            if (plan.creditGainPeriodSeconds > 0) {
                const creditsGainPerHour = Math.round(
                    plan.creditGainAmount / plan.creditGainPeriodSeconds * 3600 * 10
                ) / 10;
                response += `📈 Восполнение: ${creditsGainPerHour} к/ч\n`;
            }
            if (user.usagePlanExpiry) {
                const expiry = user.usagePlanExpiry;
                const expiryDateString = `${expiry.getDate()}.${expiry.getMonth() + 1}.${expiry.getFullYear()}`;
                response += `📆 Действует до ${expiryDateString}\n`;
            } else {
                response += `♾️ Действует вечно\n`;
            }
        }
        await vkMessagesService.send(message.peerId, response);
    }

    private async handleList(message: VkMessage): Promise<void> {
        const { vkMessagesService } = this.context;
        let plans = await this.context.usagePlanOrmService.getUsagePlans();
        plans = plans
            .filter(plan => plan.visible)
            .sort(
                (a, b) => a.creditGainAmount/a.creditGainPeriodSeconds - b.creditGainAmount/b.creditGainPeriodSeconds
            );
        let response = `Планы:\n`;
        for (const plan of plans) {
            const creditsGainPerHour = Math.round(
                plan.creditGainAmount / plan.creditGainPeriodSeconds * 3600 * 10
            ) / 10;
            response += `- ${plan.title} (${creditsGainPerHour} к/ч, макс. ${plan.maxCredits})\n`;
        }
        await vkMessagesService.send(message.peerId, response);
    }

    private getUsage(): string {
        let result = '';
        result += 'Команды:\n';
        result += '/sweet plan\n';
        result += '/sweet plan list\n';
        result += '/sweet plan help\n';
        return result;
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/command/PlanCommand.kt
package space.davids_digital.sweetie.command

import org.springframework.stereotype.Component
import space.davids_digital.sweetie.integration.vk.VkMessagesService
import space.davids_digital.sweetie.model.VkMessageModel
import space.davids_digital.sweetie.orm.service.UsagePlanOrmService
import space.davids_digital.sweetie.orm.service.VkUserOrmService
import java.time.format.DateTimeFormatter
import kotlin.math.round

@Component
class PlanCommand(
    private val vkMessagesService: VkMessagesService,
    private val usagePlanOrmService: UsagePlanOrmService,
    private val vkUserOrmService: VkUserOrmService

): Command {
    override fun getNames(): Array<String> {
        return arrayOf("plan")
    }

    override fun getUsage(): String {
        return "plan (...)"
    }

    override fun requiresChatAdmin(): Boolean {
        return true
    }

    override fun requiresAppCeo(): Boolean {
        return true
    }

    override suspend fun handle(commandName: String, rawArguments: String, message: VkMessageModel) {
        val subCommand = rawArguments.split(" ")[0]
        when (subCommand) {
            "" -> handleDefault(message)
            "list" -> handleList(message)
            "help" -> handleHelp(message)
            else -> handleHelp(message)
        }
    }

    private fun handleDefault(message: VkMessageModel) {
        val user = vkUserOrmService.getById(message.fromId) ?: throw CommandException("Пользователь не найден")
        val plan = usagePlanOrmService.getOrDefault(user.usagePlanId, "default")
        val planTitle = plan?.title ?: "(no plan)"
        val builder = StringBuilder()
        builder.append("⚡ Sweetie AI $planTitle\n")
        builder.append("\uD83E\uDE99 Кредиты: ${user.credits}")
        if (plan != null) {
            builder.append("/${plan.maxCredits}")
        }
        builder.append("\n")
        if (plan != null) {
            if (plan.creditGainPeriodSeconds > 0) {
                val creditsGainPerHour = round(
                    plan.creditGainAmount.toDouble()/plan.creditGainPeriodSeconds * 3600 * 10
                ) / 10
                builder.append("\uD83D\uDCC8 Восполнение: $creditsGainPerHour к/ч\n")
            }
            if (user.usagePlanExpiry != null) {
                val expiryText = user.usagePlanExpiry.format(DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm"))
                builder.append("\uD83D\uDCC6 Действует до $expiryText\n")
            } else {
                builder.append("♾\uFE0F Действует вечно\n")
            }
        }
        vkMessagesService.send(message.peerId, builder.toString())
    }

    private fun handleList(message: VkMessageModel) {
        val plans = usagePlanOrmService.findAll()
            .filter { it.visible }
            .sortedBy { it.creditGainAmount.toDouble()/it.creditGainPeriodSeconds }
        val builder = StringBuilder()
        builder.append("Планы:\n")
        plans.forEach {
            val creditsGainPerHour = round(it.creditGainAmount.toDouble()/it.creditGainPeriodSeconds * 3600 * 10) / 10
            builder.append("- ${it.title} ($creditsGainPerHour к/ч, макс. ${it.maxCredits})\n")
        }
        vkMessagesService.send(message.peerId, builder.toString())
    }

    private fun handleHelp(message: VkMessageModel) {
        val text = """
            Команды:
            /sweet plan
            /sweet plan list
            /sweet plan help
        """.trimIndent()
        vkMessagesService.send(message.peerId, text)
    }
}